[server] Improve WMS GetFeatureInfo text/html output

pathmapper opened the pull request:
Closes https://github.com/qgis/QGIS/issues/49851

- improve styling via CSS
- use also layer title (before only layer name)
- if no result is found for a layer, don't show a table
- dont show internal QGIS feature id

All kudos belong to @astroidex who started this.

astroidex committed:
improved styling for featurinfo html output #49851, todo update test

pathmapper committed:
Some adjustments

pathmapper committed:
Update tests

elpaso commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -1649,6 +1649,15 @@ namespace QgsWms
             }
 
             layerElement.setAttribute( QStringLiteral( "name" ), layerName );
+            QString layerTitle = layer->title();
```
```suggestion
            const QString layerTitle = layer->title();
```

elpaso commented on src/server/services/wms/qgswmsrenderer.cpp:
```diff
@@ -2571,74 +2580,97 @@ namespace QgsWms
-          featureInfoString.append( "<TR><TH>Feature</TH><TD>" + featureElement.attribute( QStringLiteral( "id" ) ) +
-                                    "</TD></TR>\n" );
+          const QDomElement featureElement = featureNodeList.at( j ).toElement();
+          featureInfoString.append( QStringLiteral( R"HTML(<table>)HTML" ) );
```
Why are you using 'R'aw strings? There aren't any quotes.

pathmapper replied:
They were included in the original commit and I didn't change them, so I can only guess that they are used because there are quotes in the head<br>
-> https://github.com/qgis/QGIS/pull/55751/files#diff-2b31f4f76dacb4bd2240768e67fbe59d416219b9b58b3fece7706a1ebb18fbaeR2587)<br>
and also to get a nice CSS formatting without a need to use `\n`.
 
That said, if you don't want to use 'R'aw strings I'm happy to replace them.

elpaso replied:
I am a big fan of raw strings and I wasn't complaining about raw strings when they are useful (like https://github.com/qgis/QGIS/pull/55751/files#diff-2b31f4f76dacb4bd2240768e67fbe59d416219b9b58b3fece7706a1ebb18fbaeR2583) but when they aren't, so please remove them unless necessary/useful (i.e to avoid backslash quoting of double quotes or newlines).

pathmapper replied:
Done in https://github.com/qgis/QGIS/pull/55751/commits/b9f2825204b544908641b5cbc16e595c86422344 along with HTML formatting improvements.

elpaso replied:
> Done in [b9f2825](https://github.com/qgis/QGIS/commit/b9f2825204b544908641b5cbc16e595c86422344) along with HTML formatting improvements.

Looks good!

I have just a final request: we should make sure that the code behaves correctly when HTML tag chars are used in the strings, can you please add test cases with:

1. layer name containing `<` and `>`
2. layer title containing `<` and `>`
3. attribute name  containing `<` and `>`
4. attribute value  containing `<` and `>`

please check that valid HTML is produced.

pathmapper replied:
Good point :+1: , this was an issue also before this PR :wink:

Addressed per https://github.com/qgis/QGIS/pull/55751/commits/fddb2363989f088baf79c237aefa4a27dc3083f1

I've used [toHtmlEscaped()](https://doc.qt.io/qt-5/qstring.html#toHtmlEscaped) except for the maptip attribute (if present) because HTML could be wanted here.

I think it's the responsibility of the user to provide valid content for the maptip.

elpaso replied:
> Good point 👍 , this was an issue also before this PR 😉

Yeah, I know it isn't your fault, thank you for fixing this!

> 
> Addressed per [fddb236](https://github.com/qgis/QGIS/commit/fddb2363989f088baf79c237aefa4a27dc3083f1)
> 
> I've used [toHtmlEscaped()](https://doc.qt.io/qt-5/qstring.html#toHtmlEscaped) except for the maptip attribute (if present) because HTML could be wanted here.
> 
> I think it's the responsibility of the user to provide valid content for the maptip.

Yes, that makes sense.

pathmapper replied:
Thank you too.

pathmapper committed:
Update src/server/services/wms/qgswmsrenderer.cpp

Co-authored-by: Alessandro Pasotti <elpaso@itopen.it>

pathmapper committed:
Address code review, improve HTML formatting, fix remaining test

pathmapper committed:
Produce valid content for strings with HTML tag chars

Tests added, ref https://github.com/qgis/QGIS/pull/55751#discussion_r1441459922

pathmapper committed:
Update Postgres tests

pathmapper committed:
Update more tests

pathmapper commented:
All tests are passing now.

elpaso commented on tests/testdata/qgis_server/getcapabilities.txt:
```diff
@@ -199,6 +199,54 @@ Content-Type: text/xml; charset=utf-8
     </Style>
    </Layer>
+   <Layer queryable="1">
+    <Name>&lt;test layer name></Name>
```
Is this the expected output? Shouldn't it be 
```suggestion
    <Name>&lt;test layer name&gt;</Name>
```
??

pathmapper replied:
> Is this the expected output?

In the scope of this PR, yes, because this is from the GetCapabilities output which isn't affected by `convertFeatureInfoToHtml()`.

This is also expected with `master` -> replacing only `<` with `&lt;` (and not `>` with `&gt;`) is the case for the XML when you save a QGIS project.
This is also where these changes from the GetCapabilities output are coming from, there are now two additional layers with tag chars in name/title included for `test_project.qgs` for testing `convertFeatureInfoToHtml()`.

Maybe it's enough to just replace the opening tag char for XML - since [toHtmlEscaped()](https://doc.qt.io/qt-5/qstring.html#toHtmlEscaped) is available, I've opted to use it for `convertFeatureInfoToHtml()` and thus there is now also the closing tag char replaced (and also `&` and `"`).

elpaso commented on tests/testdata/qgis_server/getcapabilities.txt:
```diff
@@ -199,6 +199,54 @@ Content-Type: text/xml; charset=utf-8
+    </Style>
+   </Layer>
+   <Layer queryable="1">
+    <Name>&lt;test layer title></Name>
```
same

pathmapper replied:
same as above

elpaso commented on tests/testdata/qgis_server/getcapabilities.txt:
```diff
@@ -199,6 +199,54 @@ Content-Type: text/xml; charset=utf-8
+   </Layer>
+   <Layer queryable="1">
+    <Name>&lt;test layer title></Name>
+    <Title>&lt;title></Title>
```
same

pathmapper replied:
same as above

elpaso commented on tests/testdata/qgis_server/getcapabilities-map.txt:
```diff
@@ -199,6 +199,54 @@ Content-Type: text/xml; charset=utf-8
+   <Layer queryable="1">
+    <Name>&lt;test layer title></Name>
+    <Title>&lt;title></Title>
+    <Abstract>A test vector layer with tag chars with title</Abstract>
```
```suggestion
    <Abstract>A test vector layer with &lt;tag&gt; chars with title</Abstract>
```
maybe test abstract too...

pathmapper replied:
Done in https://github.com/qgis/QGIS/pull/55751/commits/4763dd017071b770eb93991f9ba0161d3bca2ca4 (as outlined above, when saving a QGIS project with `<` `>`, only the opening tag char is replaced for the XML).

pathmapper committed:
Test also layer server Abstract

ref https://github.com/qgis/QGIS/pull/55751#pullrequestreview-1807328965
